src/main/java/com/gotocompany/depot/maxcompute/client/insert/InsertManager.java

```diff
@@ -0,0 +1,48 @@
+package com.gotocompany.depot.maxcompute.client.insert;
+
+import com.aliyun.odps.tunnel.TableTunnel;
+import com.aliyun.odps.tunnel.TunnelException;
+import com.aliyun.odps.tunnel.io.CompressOption;
+import com.gotocompany.depot.config.MaxComputeSinkConfig;
+import com.gotocompany.depot.maxcompute.model.RecordWrapper;
+import com.gotocompany.depot.metrics.Instrumentation;
+import com.gotocompany.depot.metrics.MaxComputeMetrics;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.List;
+
+@RequiredArgsConstructor
+@Getter
+public abstract class InsertManager {
+
+    private final TableTunnel tableTunnel;
+    private final MaxComputeSinkConfig maxComputeSinkConfig;
+    private final Instrumentation instrumentation;
+    private final MaxComputeMetrics maxComputeMetrics;
+
+    public abstract void insert(List<RecordWrapper> recordWrappers) throws TunnelException, IOException;
+
+    protected TableTunnel.StreamRecordPack newRecordPack(TableTunnel.StreamUploadSession streamUploadSession) throws IOException, TunnelException {
+        if (!maxComputeSinkConfig.isStreamingInsertCompressEnabled()) {
+            return streamUploadSession.newRecordPack();
+        }
+        return streamUploadSession.newRecordPack(new CompressOption(maxComputeSinkConfig.getMaxComputeCompressionAlgorithm(),
+                maxComputeSinkConfig.getMaxComputeCompressionLevel(),
+                maxComputeSinkConfig.getMaxComputeCompressionStrategy()));
+    }
+
+    protected void instrument(Instant start, TableTunnel.FlushResult flushResult) {
+        instrumentation.incrementCounter(maxComputeMetrics.getMaxComputeOperationTotalMetric(),
+                String.format(MaxComputeMetrics.MAXCOMPUTE_API_TAG, MaxComputeMetrics.MaxComputeAPIType.TABLE_INSERT));
+        instrumentation.captureDurationSince(maxComputeMetrics.getMaxComputeOperationLatencyMetric(), start,
+                String.format(MaxComputeMetrics.MAXCOMPUTE_API_TAG, MaxComputeMetrics.MaxComputeAPIType.TABLE_INSERT));
```

Comment:
The instrumentation logic is tightly coupled with the InsertManager. Consider extracting this into a separate MetricsCollector class following single responsibility principle. This would improve testability and make the metrics collection more flexible.

Reply:
Let's do this for all the sink then

Reply:
It seems putting the instrumentation logic within the InsertManager makes thing more cohesive. Since the metrics collected also not generic and tightly related to insertion process.